How to use it

Hi, 
I fill very stupid but ... I followed the issues on pydantic I end-up here, I installed it,  but I have no idea how to use it ! 

What should I do with the  `eval_type_backport`  to make it work with pydantic ? 
I though I could replace  `typing._eval_type` method to this one by I end up with cycling calls. 

Could you drop a few lines in the Readme ? 

cheers, 
Sylvain

Huh actually I'm still not managing to use it correctly :thinking: 

With

```python
eval_type(  # noqa: PGH001,S307
    param.annotation,
    exec_globals,
    {},
    try_default=False,
)
```

...`param.annotation` being `"str | None"`

```pytb
    def _eval_direct(
        value: typing.ForwardRef,
        globalns: dict[str, Any] | None = None,
        localns: dict[str, Any] | None = None,
    ):
>       tree = ast.parse(value.__forward_arg__, mode='eval')
E       AttributeError: 'str' object has no attribute '__forward_arg__'
```

I suppose I should manually wrap any string annotation into a forward ref?

--- a/README.md
+++ b/README.md
@@ -2,13 +2,34 @@
 
 [![Build Status](https://github.com/alexmojaki/eval_type_backport/workflows/Tests/badge.svg)](https://github.com/alexmojaki/eval_type_backport/actions) [![Coverage Status](https://coveralls.io/repos/github/alexmojaki/eval_type_backport/badge.svg)](https://coveralls.io/github/alexmojaki/eval_type_backport) [![Supports Python versions 3.7+, including PyPy](https://img.shields.io/pypi/pyversions/eval_type_backport.svg)](https://pypi.python.org/pypi/eval_type_backport) [![Anaconda's conda-forge channel](https://anaconda.org/conda-forge/eval-type-backport/badges/version.svg)](https://anaconda.org/conda-forge/eval-type-backport)
 
-This is a tiny package providing a replacement for `typing._eval_type` to support newer typing features in older Python versions.
-
-Yes, that's very specific, and yes, `typing._eval_type` is a protected function that you shouldn't normally be using. Really this package is specifically made for https://github.com/pydantic/pydantic/issues/7873.
-
+This package makes runtime typing inspection with the `typing` module possible with newer syntax in older Python versions.
 Specifically, this transforms `X | Y` into `typing.Union[X, Y]`
 and `list[X]` into `typing.List[X]` etc. (for all the types made generic in PEP 585)
 if the original syntax is not supported in the current Python version.
+
+For users of Pydantic, merely having this package installed should make Pydantic models with newer syntax work in older Python versions.
+
+Here's an example of how to use it directly:
+
+```python
+import typing
+
+from eval_type_backport import install_patch
+
+install_patch()
+
+
+class Foo:
+    a: 'int | str'
+
+
+print(typing.get_type_hints(Foo))
+```
+
+In Python 3.9, without the `install_patch`, the above code will raise `TypeError: unsupported operand type(s) for |: 'type' and 'type'`.
+With it, it prints `{'a': typing.Union[int, str]}` as expected.
+
+`install_patch` monkey-patches the `typing` module so that things should 'just work' from wherever. To specifically use the backport where you need it instead of patching globally, you can use the `eval_type_backport.eval_type_backport` function directly instead of `typing._eval_type`. Yes, `typing._eval_type` is a protected function that you shouldn't normally be using. This package was written to support Pydantic, which uses it internally.
 
 ## Install
 

--- a/eval_type_backport/__init__.py
+++ b/eval_type_backport/__init__.py
@@ -1,4 +1,4 @@
-from .eval_type_backport import ForwardRef, eval_type_backport
+from .eval_type_backport import ForwardRef, eval_type_backport, install_patch
 
 try:
     from .version import __version__
@@ -6,4 +6,4 @@ except ImportError:  # pragma: no cover
     # version.py is auto-generated with the git tag when building
     __version__ = '???'
 
-__all__ = ['ForwardRef', 'eval_type_backport', '__version__']
+__all__ = ['install_patch', 'ForwardRef', 'eval_type_backport', '__version__']

--- a/eval_type_backport/eval_type_backport.py
+++ b/eval_type_backport/eval_type_backport.py
@@ -215,7 +215,8 @@ def _eval_direct(
 
 
 if sys.version_info[:2] >= (3, 10):
-    def eval_type_backport(
+
+    def eval_type_backport(  # type: ignore  # allow duplicate declaration
         value: Any,
         globalns: dict[str, Any] | None = None,
         localns: Mapping[str, Any] | None = None,

--- a/eval_type_backport/eval_type_backport.py
+++ b/eval_type_backport/eval_type_backport.py
@@ -9,10 +9,7 @@ import sys
 import typing
 import uuid
 from collections.abc import Mapping
-from typing import TYPE_CHECKING, Any
-
-if TYPE_CHECKING:  # pragma: no cover
-    from _typeshed import Unused
+from typing import Any
 
 
 def is_unsupported_types_for_union_error(e: TypeError) -> bool:
@@ -214,41 +211,39 @@ def _eval_direct(
     return transformer.eval_type(tree, original_ref=value)
 
 
-if sys.version_info[:2] >= (3, 10):
-    def eval_type_backport(
-        value: Any,
-        globalns: dict[str, Any] | None = None,
-        localns: Mapping[str, Any] | None = None,
-        try_default: Unused = True,
-        *args: Any,
-        **kwargs: Any,
-    ) -> Any:
-        """Alias to typing._eval_type (Python 3.10+)."""
+def eval_type_backport(
+    value: Any,
+    globalns: dict[str, Any] | None = None,
+    localns: Mapping[str, Any] | None = None,
+    try_default: bool = True,
+    *args: Any,
+    **kwargs: Any,
+) -> Any:
+    """
+    Like `typing._eval_type`, but lets older Python versions use newer typing features.
+    Specifically, this transforms `X | Y` into `typing.Union[X, Y]`
+    and `list[X]` into `typing.List[X]` etc. (for all the types made generic in PEP 585)
+    if the original syntax is not supported in the current Python version.
+    """
+    if sys.version_info[:2] >= (3, 10):
         return typing._eval_type(value, globalns, localns, *args, **kwargs)  # type: ignore
 
-else:
+    if not try_default:
+        return _eval_direct(value, globalns, localns)
+    try:
+        return typing._eval_type(  # type: ignore
+            value, globalns, localns, *args, **kwargs
+        )
+    except TypeError as e:
+        if not (isinstance(value, typing.ForwardRef) and is_backport_fixable_error(e)):
+            raise
+        return _eval_direct(value, globalns, localns)
 
-    def eval_type_backport(
-        value: Any,
-        globalns: dict[str, Any] | None = None,
-        localns: Mapping[str, Any] | None = None,
-        try_default: bool = True,
-    ) -> Any:
-        """
-        Like `typing._eval_type`, but lets older Python versions use newer typing features.
-        Specifically, this transforms `X | Y` into `typing.Union[X, Y]`
-        and `list[X]` into `typing.List[X]` etc. (for all the types made generic in PEP 585)
-        if the original syntax is not supported in the current Python version.
-        """
-        if not try_default:
-            return _eval_direct(value, globalns, localns)
-        try:
-            return typing._eval_type(  # type: ignore
-                value, globalns, localns
-            )
-        except TypeError as e:
-            if not (
-                isinstance(value, typing.ForwardRef) and is_backport_fixable_error(e)
-            ):
-                raise
-            return _eval_direct(value, globalns, localns)
+
+def install_patch() -> None:
+    """Monkey-patch `typing.ForwardRef._evaluate` to support newer syntax on older Python versions.
+
+    This indirectly makes functions like `typing.get_type_hints` and `typing._eval_type` work as well.
+    """
+    if sys.version_info[:2] < (3, 10):
+        typing.ForwardRef._evaluate = ForwardRef._evaluate  # type: ignore

--- a/eval_type_backport/eval_type_backport.py
+++ b/eval_type_backport/eval_type_backport.py
@@ -9,7 +9,10 @@ import sys
 import typing
 import uuid
 from collections.abc import Mapping
-from typing import Any
+from typing import TYPE_CHECKING, Any
+
+if TYPE_CHECKING:  # pragma: no cover
+    from _typeshed import Unused
 
 
 def is_unsupported_types_for_union_error(e: TypeError) -> bool:
@@ -211,33 +214,44 @@ def _eval_direct(
     return transformer.eval_type(tree, original_ref=value)
 
 
-def eval_type_backport(
-    value: Any,
-    globalns: dict[str, Any] | None = None,
-    localns: Mapping[str, Any] | None = None,
-    try_default: bool = True,
-    *args: Any,
-    **kwargs: Any,
-) -> Any:
-    """
-    Like `typing._eval_type`, but lets older Python versions use newer typing features.
-    Specifically, this transforms `X | Y` into `typing.Union[X, Y]`
-    and `list[X]` into `typing.List[X]` etc. (for all the types made generic in PEP 585)
-    if the original syntax is not supported in the current Python version.
-    """
-    if sys.version_info[:2] >= (3, 10):
+if sys.version_info[:2] >= (3, 10):
+    def eval_type_backport(
+        value: Any,
+        globalns: dict[str, Any] | None = None,
+        localns: Mapping[str, Any] | None = None,
+        try_default: Unused = True,
+        *args: Any,
+        **kwargs: Any,
+    ) -> Any:
+        """Alias to typing._eval_type (Python 3.10+)."""
         return typing._eval_type(value, globalns, localns, *args, **kwargs)  # type: ignore
 
-    if not try_default:
-        return _eval_direct(value, globalns, localns)
-    try:
-        return typing._eval_type(  # type: ignore
-            value, globalns, localns, *args, **kwargs
-        )
-    except TypeError as e:
-        if not (isinstance(value, typing.ForwardRef) and is_backport_fixable_error(e)):
-            raise
-        return _eval_direct(value, globalns, localns)
+else:
+
+    def eval_type_backport(
+        value: Any,
+        globalns: dict[str, Any] | None = None,
+        localns: Mapping[str, Any] | None = None,
+        try_default: bool = True,
+    ) -> Any:
+        """
+        Like `typing._eval_type`, but lets older Python versions use newer typing features.
+        Specifically, this transforms `X | Y` into `typing.Union[X, Y]`
+        and `list[X]` into `typing.List[X]` etc. (for all the types made generic in PEP 585)
+        if the original syntax is not supported in the current Python version.
+        """
+        if not try_default:
+            return _eval_direct(value, globalns, localns)
+        try:
+            return typing._eval_type(  # type: ignore
+                value, globalns, localns
+            )
+        except TypeError as e:
+            if not (
+                isinstance(value, typing.ForwardRef) and is_backport_fixable_error(e)
+            ):
+                raise
+            return _eval_direct(value, globalns, localns)
 
 
 def install_patch() -> None:

--- a/tests/test_eval_type_backport.py
+++ b/tests/test_eval_type_backport.py
@@ -2,7 +2,6 @@ from __future__ import annotations
 
 import collections
 import contextlib
-import importlib
 import re
 import sys
 import typing as t
@@ -17,9 +16,6 @@ str((collections, contextlib, re))  # mark these as used (by eval calls)
 
 
 eval_type = t._eval_type  # type: ignore[attr-defined]
-eval_type_backport_impl = importlib.import_module(
-    'eval_type_backport.eval_type_backport'
-)
 
 
 def eval_kwargs(code: str):
@@ -385,15 +381,5 @@ def test_install_patch_supports_get_type_hints():
             assert t.ForwardRef._evaluate is ForwardRef._evaluate
         else:
             assert t.ForwardRef._evaluate is original_evaluate
-    finally:
-        t.ForwardRef._evaluate = original_evaluate
-
-
-def test_install_patch_is_noop_on_newer_python(monkeypatch):
-    original_evaluate = t.ForwardRef._evaluate
-    try:
-        monkeypatch.setattr(eval_type_backport_impl.sys, 'version_info', (3, 10))
-        install_patch()
-        assert t.ForwardRef._evaluate is original_evaluate
     finally:
         t.ForwardRef._evaluate = original_evaluate

--- a/tests/test_eval_type_backport.py
+++ b/tests/test_eval_type_backport.py
@@ -2,6 +2,7 @@ from __future__ import annotations
 
 import collections
 import contextlib
+import importlib
 import re
 import sys
 import typing as t
@@ -16,6 +17,9 @@ str((collections, contextlib, re))  # mark these as used (by eval calls)
 
 
 eval_type = t._eval_type  # type: ignore[attr-defined]
+eval_type_backport_impl = importlib.import_module(
+    'eval_type_backport.eval_type_backport'
+)
 
 
 def eval_kwargs(code: str):
@@ -381,5 +385,15 @@ def test_install_patch_supports_get_type_hints():
             assert t.ForwardRef._evaluate is ForwardRef._evaluate
         else:
             assert t.ForwardRef._evaluate is original_evaluate
+    finally:
+        t.ForwardRef._evaluate = original_evaluate
+
+
+def test_install_patch_is_noop_on_newer_python(monkeypatch):
+    original_evaluate = t.ForwardRef._evaluate
+    try:
+        monkeypatch.setattr(eval_type_backport_impl.sys, 'version_info', (3, 10))
+        install_patch()
+        assert t.ForwardRef._evaluate is original_evaluate
     finally:
         t.ForwardRef._evaluate = original_evaluate

--- a/tests/test_eval_type_backport.py
+++ b/tests/test_eval_type_backport.py
@@ -370,7 +370,7 @@ def test_install_patch_supports_get_type_hints():
     assert eval_type_backport_module.install_patch is install_patch
 
     class Foo:
-        value: 'int | str'
+        value: int | str
 
     original_evaluate = t.ForwardRef._evaluate
     try:

--- a/tests/test_eval_type_backport.py
+++ b/tests/test_eval_type_backport.py
@@ -8,7 +8,8 @@ import typing as t
 
 import pytest
 
-from eval_type_backport import ForwardRef, eval_type_backport
+import eval_type_backport as eval_type_backport_module
+from eval_type_backport import ForwardRef, eval_type_backport, install_patch
 from eval_type_backport.eval_type_backport import new_generic_types
 
 str((collections, contextlib, re))  # mark these as used (by eval calls)
@@ -362,3 +363,23 @@ def test_copy_forward_ref_attrs():
         **({} if sys.version_info < (3, 9, 8) else {'is_class': True}),
     )
     eval_type_backport(ref, globalns=globals(), localns=locals())
+
+
+def test_install_patch_supports_get_type_hints():
+    assert 'install_patch' in eval_type_backport_module.__all__
+    assert eval_type_backport_module.install_patch is install_patch
+
+    class Foo:
+        pass
+
+    Foo.__annotations__ = {'value': 'int | str'}
+    original_evaluate = t.ForwardRef._evaluate
+    try:
+        install_patch()
+        assert t.get_type_hints(Foo) == {'value': t.Union[int, str]}
+        if sys.version_info[:2] < (3, 10):
+            assert t.ForwardRef._evaluate is ForwardRef._evaluate
+        else:
+            assert t.ForwardRef._evaluate is original_evaluate
+    finally:
+        t.ForwardRef._evaluate = original_evaluate

--- a/tests/test_eval_type_backport.py
+++ b/tests/test_eval_type_backport.py
@@ -370,9 +370,8 @@ def test_install_patch_supports_get_type_hints():
     assert eval_type_backport_module.install_patch is install_patch
 
     class Foo:
-        pass
+        value: 'int | str'
 
-    Foo.__annotations__ = {'value': 'int | str'}
     original_evaluate = t.ForwardRef._evaluate
     try:
         install_patch()

